File: 41.first-missing-positive.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public class Solution41
{
    public int FirstMissingPositive(int[] nums) {
        int minVal = int.MaxValue;
        int length =0;
        foreach(int num in nums)
        {
            if(num > 0 && num < minVal)
            {
                minVal = num;
            }
            if(num > 0)
            {
                length ++;
            }
        }

        if(minVal != 1)
        {
            return 1;
        }

        int[] numMap = new int[length];
        foreach(int num in nums)
        {
            // if the num is outsize the range of minVal_length, no need to process this value.
            if(num > 0 && num - minVal <= length -1)
            {
                numMap[num-minVal] = 1;
            }
        }
        
        for(int i =0; i< length; i++)
        {
            if(numMap[i] != 1)
                return i+1;
        }
        
        return length+1;
    }
}

File: 41_first-missing-positive.py
from typing import List


class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        size = len(nums)
        for i in range(size):
            if nums[i] == i+1 : continue
            if nums[i] < 1 or nums[i] > size: 
                nums[i] = -1
                continue
            
            target = nums[i]
            nums[i] = -1
            while target > 0 and target <= size and nums[target -1] != target:
                tmp = nums[target -1]
                nums[target -1] = target

                target = tmp

        for i in range(size):
            if nums[i] == -1:
                return i+1
        
        return size + 1

instance = Solution()
print(instance.firstMissingPositive([3,4,-1,1]))

